onnxruntime/core/providers/xnnpack/xnnpack_init.h

```diff
@@ -46,6 +46,19 @@ namespace xnnpack {
 #define XNN_ALLOCATION_ALIGNMENT 16
 #endif
 
+#if (!defined(_MSC_VER)) || (_MSC_VER >= 1930)
+#if defined(_M_ARM64) || defined(__aarch64__) || defined(_M_ARM64EC)
+#if !defined(__APPLE__)
+// Had to temporary disable fp16 under APPLE ARM64, as compiling
```

Comment:
The iOS/macOS arm64 packages include xnnpack, and having a fallback on CPU if someone has a fp16 model they're try to use with CoreML is a key scenario. 

Is this issue potentially fixed by something like this or it's a different issue?

https://github.com/Maratyszcza/FP16/pull/32/files

If not, we should limit this to when we are actually building a universal binary. I don't think that applies to the cocoapod packages as they have xcframeworks in them (different framework is built for each platform). And currently it looks like the pipeline that does the mac build with the universal binary doesn't enable xnnpack.

https://github.com/microsoft/onnxruntime/blob/1bda91fc57d61367d245db808106687b43036bd1/tools/ci_build/github/azure-pipelines/templates/mac-cpu-packing-jobs.yml#L99

Reply:
Not allowing a universal2 build with xnnpack seems to be enforced at the cmake level so any additional conditions on it being __APPLE__ don't seem to be required here.

> -- Building ONNX Runtime with XNNPACK and multiple OSX architectures is not supported. Got:(arm64;x86_64). Please specify a single architecture in CMAKE_OSX_ARCHITECTURES and re-configure.
CMake Error at external/xnnpack.cmake:47 (MESSAGE):

As XNNPACK is a different world to MLAS and we build it in different places/ways, instead of copying the MLAS conditions can we be super simple and do the following:

Copy the ARM64 #if logic from xnnpack common.h (we have logic for a number of other platforms already), and simply use that for now. If we need to make it more complicated with additional checks we can do that on an as-needed basis. 

```c++
#if defined(__aarch64__) || defined(_M_ARM64) || defined(_M_ARM64EC)
  #define XNN_ARCH_ARM64 1
#else
  #define XNN_ARCH_ARM64 0
#endif

...
// fp16 support can vary on a kernel by kernel basis. Keep it simple and limit to arm64 for now.
// e.g. XNNPACK maxpool has x64 and arm64 fp16 kernels.
#if XNN_ARCH_ARM64
#define XNNPACK_FP16_SUPPORTED
#endif
```

The XNNPACK headers have additional platform/arch checks as well to ensure kernels are only included when valid, so this top level #define seems like a general purpose 'attempt to use fp16' flag.